compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1049,6 +1049,59 @@ fn escape_string(s: &[u8]) -> String {
     }
 }
 
+#[cfg(not(windows))]
+fn escape_linker_output(s: &[u8], _flavour: LinkerFlavor) -> String {
+    escape_string(s)
+}
+
+/// If the output of the msvc linker is not UTF-8 and the host is Windows,
+/// then try to convert the string from the OEM encoding.
+#[cfg(windows)]
+fn escape_linker_output(s: &[u8], flavour: LinkerFlavor) -> String {
+    // This only applies to the actual MSVC linker.
+    if flavour != LinkerFlavor::Msvc(Lld::No) {
+        return escape_string(s);
+    }
+    match str::from_utf8(s) {
+        Ok(s) => return s.to_owned(),
+        Err(_) if s.len() <= i32::MAX as usize => {
+            use windows::Win32::Globalization::{
+                GetLocaleInfoEx, MultiByteToWideChar, CP_OEMCP, LOCALE_IUSEUTF8LEGACYOEMCP,
+                LOCALE_NAME_SYSTEM_DEFAULT, LOCALE_RETURN_NUMBER, MB_ERR_INVALID_CHARS,
+            };
+            // Get the legacy system OEM code page.
+            let code_page = unsafe {
+                let mut cp: u32 = 0;
+                // We're using the `LOCALE_RETURN_NUMBER` flag to return a u32.
+                // But the API requires us to pass the data as though it's a [u16] string.
+                let len = std::mem::size_of::<u32>() / std::mem::size_of::<u16>();
+                let data = std::slice::from_raw_parts_mut(&mut cp as *mut u32 as *mut u16, len);
+                let len_written = GetLocaleInfoEx(
+                    LOCALE_NAME_SYSTEM_DEFAULT,
+                    LOCALE_IUSEUTF8LEGACYOEMCP | LOCALE_RETURN_NUMBER,
+                    Some(data),
+                );
+                if len_written as usize == len { cp } else { CP_OEMCP }
+            };
+            // Error if the string is not valid for the expected code page.
+            let flags = MB_ERR_INVALID_CHARS;
+            // Call MultiByteToWideChar twice.
+            // First to calculate the length then to convert the string.
+            let mut len = unsafe { MultiByteToWideChar(code_page, flags, s, None) };
+            if len > 0 {
+                let mut utf16 = vec![0; len as usize];
+                len = unsafe { MultiByteToWideChar(code_page, flags, s, Some(&mut utf16)) };
+                if len > 0 {
+                    return String::from_utf16_lossy(&utf16[..len as usize]);
+                }
+            }
```

Comment:
Sure! I just need to find a place to put it...

Reply:
Right alongside the code here is fine. It doesn't need to be cleverly arranged, as it can be moved around later if we need to move it. It just probably should be its own thing so that it is in fact easy to move.

Reply:
Ok, I've abstracted the windows parts into a `#[cfg(windows)]` submodule just to keep things neat for now. I made it two functions, one for getting the code page and one for doing the conversion so it's more usefully abstracted. Maybe other tools use other code pages, there are a few.

I was getting hung up on bikeshedding the function name so I just gave up and picked one 🙂.

Reply:
Oh... also I did thing of moving the `from_utf8` into it but I thought that might be too weird to not be prominently opted into? I dunno though.